### ISSUE TITLE
when will top-k off-policy be implemented? 

### ISSUE THREAD

**Issue by joseph-chan** [awarebayes/RecNN#3]
when will top-k off-policy be implemented?  I was reading this paper and looking forward to its implementation released, haha

**Comment by awarebayes** [comment]
I am focusing on making the library usable/writing the docs right now. Once the package is in it's "mature" state, I will be adding new algorithms. You know, changing one function and rewriting it in ~5 files has never been cool. I d assume it would take a couple of weeks

**Comment by awarebayes** [comment]
+ reinforce TopK works on top of existing working recommenders (DDPG, TD3, SAC, BCQ) and I am still working on these. Gonna keep your issue open till it's released.

**Comment by joseph-chan** [comment]
> I am focusing on making the library usable/writing the docs right now. Once the package is in it's "mature" state, I will be adding new algorithms. You know, changing one function and rewriting it in ~5 files has never been cool. I d assume it would take a couple of weeks

It's  definitively a great job, and many people will benefit from it, thanks for your great effort and I  will pay attention to your cool work

**Comment by tangbotony** [comment]
Attention...

**Comment by jdxyw** [comment]
look forward to it

**Comment by awarebayes** [comment]
Yes, I have already started to write an article. The article is scheduled this week. Dunno about the implementation, but it is coming soon. Already started working on the implementation, got a rough sketch here: https://gist.github.com/awarebayes/9c8a0f28e83a723549ce4ed0c4731997. In the article covered most of the paper already. Make sure to follow and not miss it: https://medium.com/@awarebayes. The code will be published here. 
![carbon (6)](https://user-images.githubusercontent.com/42784580/69084633-b7c72100-0a55-11ea-8a7d-b3055a47e8dc.png)

**Comment by awarebayes** [comment]
Just implemented OffPolicy correction, without Top K yet. But TopK is coming....
[Link to the notebook](https://github.com/awarebayes/RecNN/blob/master/examples/2.%20REINFORCE%20TopK%20Off%20Policy%20Correction/2.%20Reinforce%20Off%20Policy%20Correction.ipynb
)

**Comment by joseph-chan** [comment]
> Just implemented OffPolicy correction, without Top K yet. But TopK is coming....
> [Link to the notebook](https://github.com/awarebayes/RecNN/blob/master/examples/2.%20REINFORCE%20TopK%20Off%20Policy%20Correction/2.%20Reinforce%20Off%20Policy%20Correction.ipynb)

Great job! I am waiting for your implementation, and close this issue, thank you again for your great job

**Comment by awarebayes** [comment]
TL;DR: **will keep it open until the article is published**
Just released Reinforce TopK! Here is the private article draft (not released yet): [link](https://medium.com/@awarebayes/reinforce-topk-off-policy-correction-e34381dceef8). I was gonna send it to google folks to take a quick look at, so it is somewhat official. Also, set up all of the notebooks online. 

Anyways, here is a quick look at policy loss with TopK Correction
![Screenshot](https://user-images.githubusercontent.com/42784580/69475852-5dc2b480-0de3-11ea-8257-1597658be245.png)

Without any correction:
![Screenshot](https://user-images.githubusercontent.com/42784580/69475873-ada17b80-0de3-11ea-9e72-1cfba15ccecf.png)

**Comment by joseph-chan** [comment]
Code and detailed documentation already given in the previous reply of awarebayes, I closed the issue
